ocaml/testsuite/tests/mixed-blocks/constructor_args.ml

```diff
@@ -0,0 +1,337 @@
+      x1 (Float_u.to_float x2) (Float_u.to_float x3)
+  | Ext_mixed4 (x1, x2, x3) ->
+    sprintf "Ext_mixed4 (%f, %f, %i)"
+      x1 (Float_u.to_float x2) x3
+  | Ext_mixed5 (x1, x2, x3, x4) ->
+    sprintf "Ext_mixed5 (%f, %f, %i, %f)"
+      x1 (Float_u.to_float x2) x3 (Float_u.to_float x4)
+  | Ext_mixed6 (x1, x2, x3) ->
+    sprintf "Ext_mixed6 (%f, %i, %f)"
+      x1 x2 (Float_u.to_float x3)
+  | Ext_mixed7 (x1, x2, x3, x4) ->
+    sprintf "Ext_mixed7 (%f, %i, %f, %i)"
+      x1 x2 (Float_u.to_float x3) x4
+  | Ext_mixed8 (x1, x2, x3, x4, x5) ->
+    sprintf "Ext_mixed8 (%f, %i, %f, %i, %f)"
+      x1 x2 (Float_u.to_float x3) x4 (Float_u.to_float x5)
+  | _ -> "<ext>"
+
+let print t = print_endline ("  " ^ to_string t)
+let print_ext t = print_endline ("  " ^ ext_to_string t)
+
+(**********************)
+(* Test: construction *)
+
+let toplevel = Mixed1 #7.0
+let toplevel_ext = Ext_mixed1 #8.0
+
+let run f =
+  print_endline "Test (construction)";
+  let lit = Mixed2 (3.0, #4.5) in
+  let half_lit = Mixed3 (6.0, f, #5.0) in
+  print toplevel;
+  print_ext toplevel_ext;
+  print lit;
+  print half_lit;
+;;
+
+let () = run #17.0
+
+(**********************************************************)
+(* Test: construction and destruction of the same value, to
+   exercise an optimization code path.
+*)
+
+let sum uf uf' f f' i i' =
+  Float_u.to_float uf +. Float_u.to_float uf' +. f +. f' +.
+  Float.of_int i +. Float.of_int i'
+
+let construct_and_destruct uf uf' f f' i i' =
+  let Constant = Constant in
+  let Uniform1 f = Uniform1 f in
+  let Mixed1 uf = Mixed1 uf in
+  let Mixed2 (f, uf) = Mixed2 (f, uf) in
+  let Mixed3 (f, uf, uf') = Mixed3 (f, uf, uf') in
+  let Mixed4 (f, uf, i) = Mixed4 (f, uf, i) in
+  let Mixed5 (f, uf, i, uf') = Mixed5 (f, uf, i, uf') in
+  let Mixed6 (f, i, uf) = Mixed6 (f, i, uf) in
+  let Mixed7 (f, i, uf, i') = Mixed7 (f, i, uf, i') in
+  let Mixed8 (f, i, uf, i', uf') = Mixed8 (f, i, uf, i', uf') in
+  let Ext_mixed1 uf = Ext_mixed1 uf in
+  let Ext_mixed2 (f, uf) = Ext_mixed2 (f, uf) in
+  let Ext_mixed3 (f, uf, uf') = Ext_mixed3 (f, uf, uf') in
+  let Ext_mixed4 (f, uf, i) = Ext_mixed4 (f, uf, i) in
+  let Ext_mixed5 (f, uf, i, uf') = Ext_mixed5 (f, uf, i, uf') in
+  let Ext_mixed6 (f, i, uf) = Ext_mixed6 (f, i, uf) in
+  let Ext_mixed7 (f, i, uf, i') = Ext_mixed7 (f, i, uf, i') in
+  let Ext_mixed8 (f, i, uf, i', uf') = Ext_mixed8 (f, i, uf, i', uf') in
+  let Uniform2 (f, f', i) = Uniform2 (f, f', i) in
+  sum uf uf' f f' i i'
+[@@ocaml.warning "-partial-match"]
+
+let () =
+  let uf = #5.0
+  and uf' = #5.1
+  and f = 14.2
+  and f' = 15.4
+  and i = 10
+  and i' = 12
+  in
+  let () =
+    let sum1 = sum uf uf' f f' i i' in
+    let sum2 = construct_and_destruct uf uf' f f' i i' in
+    Printf.printf
+      "Test (construct and destruct): %f = %f (%s)\n"
+      sum1
+      sum2
+      (if Float.equal sum1 sum2 then "PASS" else "FAIL")
+  in
+  ()
+
+(**************************)
+(* Test: recursive groups *)
```

Comment:
I don't think the error message has anything to do with this PR.  I can also write:

```ocaml
type t = C of float * int

let _ =
  let rec x = C (42.0, x); 3 in
  ();;
```

and get the same crash. #2394 fixes the crash for all cases, both my example and yours.

I do think that the acceptance of this program is a problem, though. Compilation fills in a dummy int `0xbbbb` for the value under construction, which is wrong for mixed blocks. Your program doesn't segfault (post #2394) but I suspect that may just be because the middle-end notices that the constructed block is local to the expression and optimizes it out.

I'll change the rec-check as you suggest.

Reply:
That's done, with an appropriate test.